enclave_wrangler/objects_api.py

```diff
@@ -1024,6 +1045,80 @@ def get_projects(verbose=False) -> List[Dict]:
+    """Will set any add any fields that don't exist on object and set them to NULL.
+
+    todo: could be nice to have a generalized function for other object types"""
+    # noinspection PyTypeChecker doesnt_understand_union_w_list_of_lists
+    obj_data_types: Dict[str, str] = cset_obj_field_datatypes(schema)
+    obj_fields: Set[str] = set(obj_data_types.keys())
+    # TODO: update it so that it sets to None for everything except numeric type, in which case it should set to 0
+    #   - did my modification below to use PG_DATATYPES_BY_GROUP work? run it
+    return [
+        cset | {
+            f: 0 if obj_data_types[f] in PG_DATATYPES_BY_GROUP['numeric'] else None
+            for f in obj_fields.difference(cset.keys())}
+        for cset in cset_objs
+    ]
+
+
+def update_cset_metadata_from_objs(
+    cset_objs: Union[Dict, List[Dict]], con: Connection = None, set_missing_fields_to_null=False, schema=SCHEMA
+):
+    """Given 1+ OMOPConceptSet objects fetched from the Enclave, update its metadata fields in the database.
+
+    :param set_missing_fields_to_null: If True, will set any missing fields to NULL in the DB. If False, will leave them
+     untouched. This is useful if the cset_objs are the result of a fetch from the Enclave. In that case, the Enclave
+     will not return fields on objects if they are NULL in the Enclave. However, if using this function to update a set
+     of specific values, and cset_objs are not the result of a fetch from the Enclave, then set this to False.
+
+    todo: Ideally would check if there were any changes before updating the DB, and do nothing if no changes.
+    todo: refactor to do all csets in 1 audit query each, and 1 update_objects_in_db() call
+    todo: performance: Ideally would update all csets with a single query, but right now stability is more important. We
+     want to update one at a time, and as soon as a single case fails, we want to throw an err message at that time.
+     This applies to the 'for cset in cset_objs' part. In order to run update_objects_in_db() on all cset_objs at once,
+     I should also update the audit_query to run on multiple objs at once.
+    """
+    conn = con if con else get_db_connection()
+    cset_objs = cset_objs if isinstance(cset_objs, list) else [cset_objs]
+
+    if set_missing_fields_to_null:
+        cset_objs = cset_objs_set_missing_fields_to_null(cset_objs, schema)
+
+    audit_query = """INSERT INTO code_sets_audit
+        SELECT *, CURRENT_TIMESTAMP AS update_timestamp
+        FROM code_sets
+        WHERE codeset_id = cset_id;"""
+    for cset in cset_objs:
+        cset_id = str(cset['codesetId'])
+        run_sql(con, audit_query.replace('cset_id', cset_id))  # backup to audit table: original
+        update_objects_in_db(con, 'OMOPConceptSet', [cset])
+        run_sql(con, audit_query.replace('cset_id', cset_id))  # backup to audit table: updated
+    if not con:
+        conn.close()
+
+
+def clean_improper_finalized_draft_metadata():
```

Comment:
## `clean_improper_finalized_draft_metadata()`
In the `objects_api` CLI.

Identify finalized drafts with improper metadata and update their metadata, title and all.